src/blog/system-analyzer.md

```diff
@@ -0,0 +1,206 @@
+  dotProduct = a.dotProduct(b);
+}
+console.log(performance.now() - start);
+```
+
+Once we run the program we notice a performance regression. Even though we are measuring the performance of two similar snippets; accessing the properties `x` and `y` of `Point` object instances by calling the `dotProduct` function in a for-loop.
+
+Snippet 1 runs approximately 3 times faster than snippet 2. The only difference being that we use negative values for `x` and `y` properties in the `Point` object in snippet 2.
+
+![Performance analysis of snippets.](/_img/system-analyzer/initial-program-performance.png)
+
+To analyse this performance difference we can use various logging options that come with V8. This is where the system analyzer shines. It can display log events and link them together with map events letting us explore the magic that is hidden within V8.
+
+Before diving more into the case study, let’s get familiar with the panels of the system analyzer tool. The tool has four main panels:
+
+- a Timeline panel to analyse Map/ICs events across time,
+- a Map panel to visualise the transition trees of the maps,
+- an IC panel to get statistics about the IC events,
+- a Source panel to display Map/IC file positions on a script.
+
+![System Analyzer Overview](/_img/system-analyzer/system-analyzer-overview.png)
+
+![Group IC events by function name to get in depth information about the IC events associated with the `dotProduct`.](/_img/system-analyzer/case1_1.png)
+
+We are analyzing how the function `dotProduct` might be causing this performance difference. So we group IC events by functionName to get more in depth information about IC events associated with the `dotProduct` function.
+
+The first thing we notice is that we have two different IC state transitions recorded by the IC events in this function. One going from uninitialised to monomorphic and the other one going from monomorphic to polymorphic. As mentioned in the previously linked [blog post](https://mathiasbynens.be/notes/shapes-ics), polymorphic IC state indicates that now we are tracking more than one Map associated with `Point` objects.
+
+We want to know why we are creating multiple Map shapes for the same type of objects we toggle the info button about IC state to get more information about the Map addresses going from uninitialised to monomorphic.
+
+![The map transition tree associated with the monomorphic IC state.](/_img/system-analyzer/case1_2.png)
+
+![The map transition tree associated with the polymorphic IC state.](/_img/system-analyzer/case1_3.png)
+
+For the monomorphic IC state we can visualise the transition tree and see that we are only dynamically adding two properties `x`and `y` but when it comes to polymorphic IC state, we have a new Map containing three properties `isNegative`, `x` and `y`.
+
+![The Map panel communicates the file position information to highlight file positions on the Source panel.](/_img/system-analyzer/case1_4.png)
```

Comment:
The text in this image: "cause dynamically adding property" -> "causes the addition of a dynamically added property"

Reply:
Images can't be edited anymore :(

Gonna leave this is as such as I don't want to delete it